### PR TITLE
 #145 Implement flag to force fragments handler to succeed

### LINKED ISSUE #145: Fragments Handler always succeed when debug mode
**Is your feature request related to a problem? Please describe.**
When debug mode is enabled and an additional header is configured (e.g `knotx-allow-invalid-fragments`), then Fragments Handler should always be successful (even some fragments have a failure status).

**Additional context**
We should be able to validate errors in [Knot.x Chrome Extension](https://github.com/Knotx/knotx-fragments-chrome-extension).

### PR DESCRIPTION
**Pull request by Voycawojka**
## Description
Fixes #145 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the [code style](/CONTRIBUTING.md#coding-conventions) of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

---
I hereby agree to the terms of the Knot.x Contributor License Agreement.

### CODE REVIEW

File: handler/core/src/main/java/io/knotx/fragments/handler/FragmentsHandler.java [hunk 1]
```diff
@@ -131,4 +131,12 @@ private RequestEvent copyRequestEvent(RequestEvent requestEvent) {
         .map(FragmentEvent::getFragment)
         .collect(Collectors.toList());
   }
+
+  private boolean shouldPassInvalidFragments(RequestContext requestContext) {
+    ClientRequest request = requestContext.getRequestEvent().getClientRequest();
+    String param = request.getParams().get(handlerOptions.getAllowInvalidFragmentsParam());
+    String header = request.getHeaders().get(handlerOptions.getAllowInvalidFragmentsHeader());
+
+    return "true".equals(param) || "true".equals(header);
```
Comment (marcinus) [inline handler/core/src/main/java/io/knotx/fragments/handler/FragmentsHandler.java:140]: How about `Boolean.parseBoolean(param)`?

File: handler/core/src/main/java/io/knotx/fragments/handler/FragmentsHandler.java [hunk 2]
```diff
@@ -36,6 +36,7 @@
 import io.vertx.core.logging.LoggerFactory;
 import io.vertx.reactivex.core.Vertx;
 import io.vertx.reactivex.ext.web.RoutingContext;
+import org.apache.commons.lang3.StringUtils;
```
Comment (marcinus) [inline handler/core/src/main/java/io/knotx/fragments/handler/FragmentsHandler.java:39]: Unused import

File: handler/core/src/main/java/io/knotx/fragments/handler/FragmentsHandler.java [hunk 3]
```diff
@@ -100,10 +101,9 @@ private void putFragments(RoutingContext routingContext, List<FragmentEvent> eve
   private RequestEventHandlerResult toHandlerResult(List<FragmentEvent> events,
       RequestContext requestContext) {
 
-    List<Fragment> failedFragments = retrieveFragments(events,
-        e -> e.getStatus() == Status.FAILURE);
+    List<Fragment> failedFragments = retrieveFragments(events, e -> e.getStatus() == Status.FAILURE);
```
Comment (marcinus) [inline handler/core/src/main/java/io/knotx/fragments/handler/FragmentsHandler.java:104]: Autoformatter wraps this line - please revert this

File: handler/core/src/main/java/io/knotx/fragments/handler/FragmentsHandler.java [hunk 4]
```diff
@@ -50,15 +51,15 @@
...
 
   FragmentsHandler(Vertx vertx, JsonObject options) {
-    FragmentsHandlerOptions handlerOptions = new FragmentsHandlerOptions(options);
+    handlerOptions = new FragmentsHandlerOptions(options);
     taskProvider = new TaskProvider(handlerOptions.getTaskFactories(), vertx);
     engine = new FragmentsEngine(vertx);
     requestContextEngine = new DefaultRequestContextEngine(getClass().getSimpleName());
-    consumerNotifier = new FragmentExecutionLogConsumersNotifier(
-        handlerOptions.getConsumerFactories());
+    consumerNotifier = new FragmentExecutionLogConsumersNotifier(handlerOptions.getConsumerFactories());
```
Comment (marcinus) [inline handler/core/src/main/java/io/knotx/fragments/handler/FragmentsHandler.java:62]: Autoformatter wraps this line - please revert this

Comment (Voycawojka) [reply handler/core/src/main/java/io/knotx/fragments/handler/FragmentsHandler.java:62]: Sure

Btw, what autoformatter? If you mean IntelliJ then it doesn't for me

File: handler/core/src/main/java/io/knotx/fragments/handler/FragmentsHandler.java [hunk 5]
```diff
@@ -98,12 +99,12 @@ private void putFragments(RoutingContext routingContext, List<FragmentEvent> eve
   }
 
   private RequestEventHandlerResult toHandlerResult(List<FragmentEvent> events,
-      RequestContext requestContext) {
+                                                    RequestContext requestContext) {
```
Comment (tomaszmichalak) [inline handler/core/src/main/java/io/knotx/fragments/handler/FragmentsHandler.java:102]: Please check your code formatter: https://github.com/Knotx/knotx/blob/master/CONTRIBUTING.md#coding-conventions

File: handler/core/src/main/java/io/knotx/fragments/handler/FragmentsHandler.java [hunk 6]
```diff
@@ -125,10 +126,18 @@ private RequestEvent copyRequestEvent(RequestEvent requestEvent) {
...
-      Predicate<FragmentEvent> predicate) {
+                                           Predicate<FragmentEvent> predicate) {
     return events.stream()
         .filter(predicate)
         .map(FragmentEvent::getFragment)
         .collect(Collectors.toList());
   }
+
+  private boolean shouldPassInvalidFragments(RequestContext requestContext) {
+    ClientRequest request = requestContext.getRequestEvent().getClientRequest();
```
Comment (tomaszmichalak) [inline handler/core/src/main/java/io/knotx/fragments/handler/FragmentsHandler.java:137]: Please provide unit tests to validate:
- no `allowInvalidFragmentsParam` is configured but passed in request
- `allowInvalidFragmentsParam` is configured and passed in request
- no `allowInvalidFragmentsHeader` is configured but passed in request
- `allowInvalidFragmentsHeader` is configured and passed in request

File: handler/core/src/main/java/io/knotx/fragments/handler/FragmentsHandlerOptions.java [hunk 7]
```diff
@@ -31,6 +31,10 @@
 
   private List<FactoryOptions> consumerFactories = Collections.emptyList();
 
+  private String allowInvalidFragmentsHeader;
```
Comment (tomaszmichalak) [inline handler/core/src/main/java/io/knotx/fragments/handler/FragmentsHandlerOptions.java:34]: Please update the documentation: https://github.com/Knotx/knotx-fragments/tree/master/handler#fragment-execution-log-consumer. Handler loads all configured consumers, then if this option is configured, then consumer gets data even if some fragments failed. This should be in this section.

**Pull request merged**